Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp

```diff
@@ -276,6 +277,41 @@ void RemoteGraphicsContextGL::simulateEventForTesting(WebCore::GraphicsContextGL
     m_context->simulateEventForTesting(event);
 }
 
+void RemoteGraphicsContextGL::getBufferSubDataInline(uint32_t target, uint64_t offset, size_t dataSize, CompletionHandler<void(std::span<const uint8_t>)>&& completionHandler)
+{
+    assertIsCurrent(workQueue());
+    static constexpr size_t getBufferSubDataInlineSizeLimit = 64 * KB; // NOTE: when changing, change the value in RemoteGraphicsContextGLProxy too.
+    if (!dataSize) {
+        completionHandler({ });
+        return;
+    }
+    MallocPtr<uint8_t> bufferStore;
+    std::span<uint8_t> bufferData;
+    if (dataSize >= getBufferSubDataInlineSizeLimit)
+        RELEASE_LOG(WebGL, "getBufferSubData requested %zu bytes of inline data", dataSize);
+    bufferStore = MallocPtr<uint8_t>::tryMalloc(dataSize);
+    if (bufferStore)
+        bufferData = { bufferStore.get(), dataSize };
+    if (!m_context->getBufferSubDataWithStatus(target, offset, bufferData))
+        bufferData = { };
+
+    completionHandler(bufferData);
+}
+
+void RemoteGraphicsContextGL::getBufferSubDataSharedMemory(uint32_t target, uint64_t offset, size_t dataSize, WebCore::SharedMemory::Handle handle, CompletionHandler<void(bool)>&& completionHandler)
+{
+    assertIsCurrent(workQueue());
+    bool validBufferData  = false;
+
+    handle.setOwnershipOfMemory(m_sharedResourceCache->resourceOwner(), WebKit::MemoryLedger::Default);
+    if (auto buffer = SharedMemory::map(WTFMove(handle), SharedMemory::Protection::ReadWrite))
+        validBufferData = m_context->getBufferSubDataWithStatus(target, offset, buffer->mutableSubspan(0, dataSize));
```

Comment:
The malicious actor might send too high dataSize, so you need to validate that the dataSize fits into the buffer.
This is, when you remove the mutableSubspan from SharedMemory.
That chescking is MESSAGE_CHECK